Android throws `ClassNotFoundException` on geoip extraction (non-fatal)

Android throws a `ClassNotFoundException` when calling `Extractor.extract` to decompress `geoip` files **if** the consumer of the `kmp-tor-binary-extract` dependency did not also add the `kmp-tor-binary-geoip` dependency. This happens because the `kmp-tor-binary-extract` module uses a `compileOnly` dependency on `kmp-tor-binary-geoip` for the android source set in order to reference its resource names `R.raw.geoip` and `R.raw.geoip6`.

**This is non-critical** as:
 - The exception is properly caught and is wrapped in `ExtractionException`, as annotated in the `extract` method's api.
 - `kmp-tor` library users are unaffected b/c `kmp-tor` has both the `extract` and `geoip` dependencies.

The error message could be improved to say something like "you should add the `kmp-tor-binary-geoip` dependency", though. Although, the only purpose of the `extract` module is to extract the `geoip` files, so it may be better to just use `implementation` so the exception is never thrown?

Need to think more on the fix for this to see if the `compileOnly` dependency could be removed entirely (which would be better).

<details>
    <summary>Stacktract</summary>

```
2022-10-23 16:05:09.914 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: io.matthewnelson.kmp.tor.manager.common.exceptions.TorManagerException: io.matthewnelson.kmp.tor.binary.extract.ExtractionException: Failed to open stream for kmptor/geoip.gz
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.TorConfigProviderAndroid.extractGeoIpV4File(TorConfigProviderAndroid.kt:60)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.manager.TorConfigProvider.retrieve$kmp_tor_manager_debug(TorConfigProvider.kt:253)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.manager.KmpTorLoader$load$validated$1.invokeSuspend(KmpTorLoader.kt:171)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.lang.Thread.run(Thread.java:920)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: Caused by: io.matthewnelson.kmp.tor.binary.extract.ExtractionException: Failed to open stream for kmptor/geoip.gz
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.internal.ExtractorDelegate.extract$kmp_tor_binary_extract_release(-ExtractorDelegate.kt:80)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.Extractor.extract(Extractor.kt:47)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.TorConfigProviderAndroid.extractGeoIpV4File(TorConfigProviderAndroid.kt:54)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: 	... 7 more
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lio/matthewnelson/kmp/tor/binary/extract/R$raw;
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.Extractor$extract$1.invoke(Extractor.kt:49)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.Extractor$extract$1.invoke(Extractor.kt:47)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.internal.ExtractorDelegate.extract$kmp_tor_binary_extract_release(-ExtractorDelegate.kt:76)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: 	... 9 more
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: Caused by: java.lang.ClassNotFoundException: Didn't find class "io.matthewnelson.kmp.tor.binary.extract.R$raw" on path: DexPathList[[zip file "/data/app/~~ThDx3_KEM-t3_oKNUzPjsw==/io.matthewnelson.kmp.tor.sample.kotlin-p8kJpGqSKF3jS1lFFNnbLw==/base.apk"],nativeLibraryDirectories=[/data/app/~~ThDx3_KEM-t3_oKNUzPjsw==/io.matthewnelson.kmp.tor.sample.kotlin-p8kJpGqSKF3jS1lFFNnbLw==/lib/x86_64, /data/app/~~ThDx3_KEM-t3_oKNUzPjsw==/io.matthewnelson.kmp.tor.sample.kotlin-p8kJpGqSKF3jS1lFFNnbLw==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:218)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: 	... 12 more
2022-10-23 16:05:09.915 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: io.matthewnelson.kmp.tor.manager.common.exceptions.TorManagerException: io.matthewnelson.kmp.tor.binary.extract.ExtractionException: Failed to open stream for kmptor/geoip6.gz
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.TorConfigProviderAndroid.extractGeoIpV6File(TorConfigProviderAndroid.kt:73)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.manager.TorConfigProvider.retrieve$kmp_tor_manager_debug(TorConfigProvider.kt:275)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.manager.KmpTorLoader$load$validated$1.invokeSuspend(KmpTorLoader.kt:171)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.lang.Thread.run(Thread.java:920)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: Caused by: io.matthewnelson.kmp.tor.binary.extract.ExtractionException: Failed to open stream for kmptor/geoip6.gz
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.internal.ExtractorDelegate.extract$kmp_tor_binary_extract_release(-ExtractorDelegate.kt:80)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.Extractor.extract(Extractor.kt:47)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.TorConfigProviderAndroid.extractGeoIpV6File(TorConfigProviderAndroid.kt:67)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: 	... 7 more
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lio/matthewnelson/kmp/tor/binary/extract/R$raw;
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.Extractor$extract$1.invoke(Extractor.kt:50)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.Extractor$extract$1.invoke(Extractor.kt:47)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at io.matthewnelson.kmp.tor.binary.extract.internal.ExtractorDelegate.extract$kmp_tor_binary_extract_release(-ExtractorDelegate.kt:76)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: 	... 9 more
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: Caused by: java.lang.ClassNotFoundException: Didn't find class "io.matthewnelson.kmp.tor.binary.extract.R$raw" on path: DexPathList[[zip file "/data/app/~~ThDx3_KEM-t3_oKNUzPjsw==/io.matthewnelson.kmp.tor.sample.kotlin-p8kJpGqSKF3jS1lFFNnbLw==/base.apk"],nativeLibraryDirectories=[/data/app/~~ThDx3_KEM-t3_oKNUzPjsw==/io.matthewnelson.kmp.tor.sample.kotlin-p8kJpGqSKF3jS1lFFNnbLw==/lib/x86_64, /data/app/~~ThDx3_KEM-t3_oKNUzPjsw==/io.matthewnelson.kmp.tor.sample.kotlin-p8kJpGqSKF3jS1lFFNnbLw==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:218)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
2022-10-23 16:05:09.916 25791-25874/io.matthewnelson.kmp.tor.sample.kotlin W/System.err: 	... 12 more
```

</details>

--- a/library/kmp-tor-binary-extract/src/androidMain/kotlin/io/matthewnelson/kmp/tor/binary/extract/Extractor.kt
+++ b/library/kmp-tor-binary-extract/src/androidMain/kotlin/io/matthewnelson/kmp/tor/binary/extract/Extractor.kt
@@ -45,15 +45,19 @@ actual class Extractor(context: Context) {
         cleanExtraction: Boolean,
     ) {
         delegate.extract(resource, destination, cleanExtraction) { resourcePath ->
-            when (resource) {
-                TorResourceGeoip -> R.raw.geoip
-                TorResourceGeoip6 -> R.raw.geoip6
-            }.let { resource ->
-                try {
-                    appContext.resources.openRawResource(resource)
-                } catch (e: Resources.NotFoundException) {
-                    throw delegate.resourceNotFound(resourcePath, e)
+            val resourceId = try {
+                when (resource) {
+                    TorResourceGeoip -> R.raw.geoip
+                    TorResourceGeoip6 -> R.raw.geoip6
                 }
+            } catch (e: ClassNotFoundException) {
+                throw ExtractionException("Dependency kmp-tor-binary-geoip is missing", e)
+            }
+
+            try {
+                appContext.resources.openRawResource(resourceId)
+            } catch (e: Resources.NotFoundException) {
+                throw delegate.resourceNotFound(resourcePath, e)
             }
         }
     }

